### ISSUE TITLE
How to deal with package trying to download and install dependency? E.g. react-dom is trying to smuggle react and fails.

### ISSUE THREAD

**Issue by cyrbon** [svanderburg/node2nix#32]
I get an error because npm is called with `--registry http://example.com` when trying to install react-dom. At first I thought that perhaps somebody forgot to change the arg value from `example.com` to a valid registry, but based on git blame, commits and comments it seems that this was introduced to prevent packages from impure downloads, so that they deliberately fail.

However, what to do when the package fails for this reason? One example is `react-dom`

To reproduce this, put `react-dom` in package-list.json, run `node2nix -i package-list.json` then run `nix-shell -E 'with import <nixpkgs> {}; with callPackage ./default.nix {}; stdenv.mkDerivation { name = "env"; src = ./.; buildInputs = [ react-dom ];  } '`.

It will fail because:

`...nodejs-4.6.0/bin/npm" "--registry" "http://www.example.com" ...`
`npm ERR! Registry returned 500 for GET on http://www.example.com/react`

In context:

```
npm WARN peerDependencies The peer dependency react included from react-dom will no
npm WARN peerDependencies longer be automatically installed to fulfill the peerDependency 
npm WARN peerDependencies in npm 3+. Your application will need to depend on it explicitly.
npm ERR! Linux 4.4.35
npm ERR! argv "/nix/store/79hy7m67b6xbj7xwwxaij6hbx79avpwv-nodejs-4.6.0/bin/node" "/nix/store/79hy7m67b6xbj7xwwxaij6hbx79avpwv-nodejs-4.6.0/bin/npm" "--registry" "http://www.example.com" "--nodedir=/nix/store/cp07nsnc1axkmnkq7173l105v2k212ym-node-sources" "--production" "install"
npm ERR! node v4.6.0
npm ERR! npm  v2.15.9
npm ERR! code E500

npm ERR! Registry returned 500 for GET on http://www.example.com/react
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /nix/store/k45vs1g175xvmfc3izj095h9hnr3wklk-node-react-dom-15.4.1/lib/node_modules/react-dom/npm-debug.log
builder for ‘/nix/store/5vfy31w7xn75n6rnwdgb0a1qpx4fjcdl-node-react-dom-15.4.1.drv’ failed with exit code 1
error: build of ‘/nix/store/5vfy31w7xn75n6rnwdgb0a1qpx4fjcdl-node-react-dom-15.4.1.drv’ failed
/run/current-system/sw/bin/nix-shell: failed to build all dependencies
```

On that note, it would be helpful for some message to be displayed, so that it's clear that it fails deliberately. I think the most simple way would be to rename `example.com` to `ImpureDownloadsAreNotAllowedAndPackageShouldNeverCallThis.com`. I can create PR for his small change, but I'm still confused about how to approach react-dom trying to download react and failing.

**Comment by cyrbon** [comment]
So, the problem is that react is not getting added as a dependency to react-dom. So, this particular problem was solved by adding react as a dependency in the generated `node-package.nix`. One need to add it to `sources` and then reference it in dependencies of `react-dom`. 

React is specified as a peer dependency in `package.json` of `react-dom` package. However, it's not getting added as a dependency during generation (after using command `node2nix -5 -i package-list.json`, but without `-5` it also fails to add it).

There are a couple of packages like this which miss some dependencies during generation and they have to be added manually. 

However, it does not really matter since those packages work only with `npm` anyway (otherwise they don't see one another, like babel plugins, etc). This includes the original package mentioned in the title - react dom - it should be installed using `npm -i` anyway. 

So, if this issue does not effect other packages, I think it can be closed.

**Comment by svanderburg** [comment]
It may still sometimes happen that there is a mismatch in what npm generates and how node2nix derives the dependency graph. Furthermore, node2nix also modifies certain version specifiers (e.g. Git URLs) to prevent npm from triggering downloads.

Unfortunately, this may cause some packages to fail even though the resulting node_modules/ structure may be the same.

The only "feature" node2nix provides to deal with such issues is to simply ignore the 'npm install' step (the README.md file describes how to do this). In most cases, this is fine since the Nix expression takes care of obtaining the dependencies.

I have also been thinking about a new strategy in which I don't replace these version specifiers and fake external processes, such as git, but this requires some deeper thought and experimentation. I have created an issue for this: https://github.com/svanderburg/node2nix/issues/26

**Comment by svanderburg** [comment]
Yesterday implemented a solution for issue #26.

However, I noticed I still run into the problem when I generate expressions for Node.js 4.x. However, good news is that if I generate expressions for Node.js 6.x, it does seem to work.

Can you try installing the development version of node2nix (see #26 for more info), and generate the expression as follows:

```bash
$ node2nix -6 -i package-list.json
```

For me it seemed to do the trick.

I haven't verified it entirely, but I suspect `react-dom` has compatibility problems with npm 2.x/node 4.x. The latest stable Node.js 6.x/npm 3.x does not seem to be affected though.
